spec/std/http/cookie_spec.cr

```diff
@@ -84,45 +86,20 @@ module HTTP
         end
       end
 
-      it "raises on invalid cookie with __Secure- prefix" do
+      it "doesn't raise on invalid cookie with __Secure- prefix" do
         cookie = HTTP::Cookie.new "x", ""
 
-        expect_raises ArgumentError, "Invalid cookie name.  Has '__Secure-' prefix, but is not secure." do
-          cookie.name = "__Secure-x"
-        end
-
-        cookie.secure = true
-
         cookie.name = "__Secure-x"
         cookie.name.should eq "__Secure-x"
+        cookie.secure.should be_false
```

Comment:
Just thinking, but what if a `__Secure` prefix automatically sets `secure = true`?

Reply:
That probably wouldn't be a bad idea.  Could probably set things up like:

* If it has a security prefix
  * and the related properties are `nil`
    * make the cookie compliant with the prefix
  * and one or more of the related properties are manually set
    * call `validate!`
      * possibly after setting the `nil` ones?

Thoughts?

Reply:
Taking another look at this. The one thing that could be unexpected is that `@secure` is typed as a `Bool` with a default value of `false`. Because of this we don't really know if the value was passed explicitly as `false` or was just left unset.

We could do one of a few things:

1. Let it be and raise an exception if you dont set `secure: true` with a `__Secure-` prefix
2. Automatically set `@secure = true` if the name has the prefix, arguably could lead to some confusion given it would be set as secure while explicitly passing `secure: false`
3. Change the constructor type/default to be `secure : Bool? = nil` and then `@secure = secure || false`. Then we could pick up on `nil` and base the auto set logic on that
4. Or did you just mean in the context of the `name=` setter?

Reply:
1. Is what's currently implemented. It's just a validation, not auto-detection.
2. would be confusing and thus bad.
3. might work
4. No, but that's a good point. The prefix should probably be validated when assigning `name` and `secure`, not just in the constructor. Currently `Cookie.new("foo", "bar").name = "__Secure-foo"` would not raise an error.

An alternative to 3. would be an additional overload where `secure` parameter is ommitted. But I think changing the default value to `nil` for autodetect seems reasonable.

Reply:
> 4\. The prefix should probably be validated when assigning `name` and `secure`, not just in the constructor.

I think I had this way back when, but we changed it to public `#valid?` and `#validate!` methods because doing that logic in the setter makes it harder to use the builder pattern to construct the cookie. Unless we're okay with needing to make the changes in a specific order. I.e. set `.secure = true` _before_ `.name = "__Secure-foo"`.

Reply:
I think I'd prefer to keep the API tied down. It's easy to forget a `validate!` call after building the instance. I suppose this would work better with a dedicated builder API such as `Cookie.build |cookie| cookie.name = "__Secure-foo"; ... end` where validity can be ensured at the end of the `build` method.

But that would probably be too much for this.

Wouldn't it be fine to implicitly set `secure = true` when assigning a name with `__Secure` prefix? If you want to go rogue you can still assign `secure = false` afterwards (and forget about `validate!`) but then you're doing that explicitly.

Reply:
Thinking a bit more extreme, what exactly is the use case for having a constructor for all the properties, while also letting the cookie be mutable via setters? Like is there actually a need for them versus just creating a cookie with the data you expect all at once? A lot of this would be resolved by just deprecating the setters and force everyone to just use the constructor. That would of course require a different issue/discussion tho.

In regards to this ticket, that approach is fine by me but my only concern is since `@secure` isn't nilable, we wouldn't know if we should make it secure or not. I.e. how to handle:

```cr
cookie.secure = false
cookie.name = "__Secure-foo"
cook.secure # => true
```

Could just be a bit confusing to explicitly set it to `false` but end up getting `true`, depending on the order you do things in. Thoughts on internally making `@secure` nilable? Can keep the API the same by having a custom getter method that would do `@secure || false`. This way we'd at least be able to pick up we shouldn't alter `@secure` since it was explicitly set at some point.

A similar thing is already possible for `__Host-` given `@path` and `@domain` are nilable already.

Reply:
Yeah, removing the setters would be a huge change. I assume they exist in order to use `Cookie` with a builder-like API.

Setting `@secure` to `nil` seems like a good idea. Although I probably wouldn't mind `#name=` to override any previous `secure` setting either.